driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java

```diff
@@ -352,8 +376,63 @@ public boolean isClosed() {
     @Override
     public <T> T sendAndReceive(final CommandMessage message, final Decoder<T> decoder, final SessionContext sessionContext,
                                 final RequestContext requestContext, final OperationContext operationContext) {
-        CommandEventSender commandEventSender;
 
+        if (!Authenticator.shouldAuthenticate(authenticator, this.description)) {
+            return sendAndReceiveInternal(message, decoder, sessionContext, requestContext, operationContext);
+        }
+        Supplier<T> sendAndReceiveInternal = () ->
+                sendAndReceiveInternal(message, decoder, sessionContext, requestContext, operationContext);
+        try {
+            return sendAndReceiveInternal.get();
+        } catch (MongoCommandException e) {
+            if (triggersReauthentication(e)) {
+                authenticated.set(false);
+                authenticator.reauthenticate(this);
+                authenticated.set(true);
+                return sendAndReceiveInternal.get();
+            }
+            throw e;
+        }
+    }
+
+
+    @Override
+    public <T> void sendAndReceiveAsync(final CommandMessage message, final Decoder<T> decoder, final SessionContext sessionContext,
+            final RequestContext requestContext, final OperationContext operationContext, final SingleResultCallback<T> callback) {
+        notNull("stream is open", stream, callback);
+
+        AsyncSupplier<T> sendAndReceiveAsyncInternal = c -> sendAndReceiveAsyncInternal(
+                message, decoder, sessionContext, requestContext, operationContext, c);
+
+        if (!Authenticator.shouldAuthenticate(authenticator, this.description)) {
+            sendAndReceiveAsyncInternal.complete(callback);
+            return;
+        }
+
+        sendAndReceiveAsyncInternal.onErrorIf(e -> triggersReauthentication(e), startAsync()
+                .run(c -> {
+                    authenticated.set(false);
+                    authenticator.reauthenticateAsync(this, c);
+                }).supply((c) -> {
+                    authenticated.set(true);
+                    sendAndReceiveAsyncInternal.complete(c);
+                }))
+        .complete(callback);
```

Comment:
Similarly to the suggestion https://github.com/mongodb/mongo-java-driver/pull/1107/files#r1214868464 in the sync PR, this code can be simplified

```suggestion
        AsyncSupplier<T> sendAndReceiveAsyncInternal = c -> sendAndReceiveAsyncInternal(
                message, decoder, sessionContext, requestContext, operationContext, c);

        sendAndReceiveAsyncInternal.onErrorSupplyIf(e ->
                        triggersReauthentication(e) && Authenticator.shouldAuthenticate(authenticator, this.description) , startAsync()
                .thenRun(c -> {
                    authenticated.set(false);
                    authenticator.reauthenticateAsync(this, c);
                }).thenSupply((c) -> {
                    authenticated.set(true);
                    sendAndReceiveAsyncInternal.complete(c);
                }))
        .complete(callback);
```